File: Demos/demo_em_scggm.py
#!/bin/python

import math
import matplotlib.pyplot as plt
import numpy as np
import scipy.sparse as ssp
import scipy.sparse.linalg as ssl
import sksparse.cholmod as skc
import sys

sys.path.append("../EM-sCGGM/")
from em_scggm import em_scggm

n = 100
n_o = 50
r = 50
q = 150
p = 200
Lambda_z = ssp.eye(r, format="coo")
Lambda_z.setdiag(0.3, 1)
Lambda_z.setdiag(0.3, -1)
Lambda_y = ssp.eye(q, format="coo")
Lambda_y.setdiag(0.3, 1)
Lambda_y.setdiag(0.3, -1)

'''
q_influence = int(math.floor(q * 0.5))
Theta_yz_top = ssp.random(q_influence, r)
Theta_yz_bottom = ssp.zeros((q-q_influence,r))
Theta_yz = ssp.vstack([Theta_yz_top, Theta_yz_bottom])
p_influence = int(math.floor(p * 0.5))
Theta_xy_top = ssp.random(p_influence, q)
Theta_xy_bottom = ssp.zeros((p-p_influence,q))
Theta_xy = ssp.vstack([Theta_xy_top, Theta_xy_bottom])
'''
Theta_yz = ssp.random(q, r)
Theta_xy = ssp.random(p, q)

X = np.random.randn(n, p)
meanY = -1*X*Theta_xy*ssl.inv(Lambda_y)
Lambda_y_factor = skc.cholesky(Lambda_y)
noiseY = (Lambda_y_factor.solve_Lt(np.random.randn(q,n))).transpose()
Y = meanY + noiseY
Y_o = Y[0:n_o,:]
meanZ = -1*Y*Theta_yz*ssl.inv(Lambda_z)
Lambda_z_factor = skc.cholesky(Lambda_z)
noiseZ = (Lambda_z_factor.solve_Lt(np.random.randn(r,n))).transpose()
Z = meanZ + noiseZ

lambdaLambda_z = 0.5
lambdaTheta_yz = 1.0
lambdaLambda_y = 0.5
lambdaTheta_xy = 1.0

(estLambda_z, estTheta_yz, estLambda_y, estTheta_xy, estStats) = em_scggm(
    Z, Y, X, lambdaLambda_z, lambdaTheta_yz, lambdaLambda_y, lambdaTheta_xy)

plt.figure()
plt.spy(Theta_xy)
plt.title("Theta")

plt.figure()
plt.spy(Lambda_y)
plt.title("Lambda")

plt.figure()
plt.spy(estLambda_y)
plt.title("estLambda")

plt.figure()
plt.spy(estTheta_xy)
plt.title("estTheta")

plt.show()

File: Demos/demo_em_scggm.m
addpath('../EM-sCGGM');
rng(0);

n = 400; n_o = 100; p = 200; q = 150; r = 50;
Lambda_y = full(spdiags([0.3*ones(q,1) ones(q,1) 0.3*ones(q,1)],[-1 0 1], q, q));
p_influence = floor(p * 0.5);
Theta_xy = [full(sprand(p_influence, q, 0.01)); zeros(p-p_influence, q)];
Lambda_z = full(spdiags([0.3*ones(r,1) ones(r,1) 0.3*ones(r,1)],[-1 0 1], r, r));
q_influence = floor(q * 0.5);
Theta_yz = [full(sprand(q_influence, r, 0.01)); zeros(q-q_influence, r)];

X = randn(n, p);
meanY = -X*Theta_xy*inv(Lambda_y); 
noiseY = ((chol(Lambda_y,'lower')')\randn(q, n))';
Y = meanY + noiseY;
meanZ = -Y*Theta_yz*inv(Lambda_z); 
noiseZ = ((chol(Lambda_z,'lower')')\randn(r, n))';
Z = meanZ + noiseZ;

Y_o = Y(1:n_o,:);

lambdaLambda_z = 0.7;
lambdaTheta_yz = 1;
lambdaLambda_y = 0.3;
lambdaTheta_xy = 0.4;
lambda.xy = lambdaTheta_xy;
lambda.yy = lambdaLambda_y;
lambda.yz = lambdaTheta_yz;
lambda.zz = lambdaLambda_z;

clear options;
options.verbose = 1;
options.sparse_E_step = 0;
options.sparse_Theta_cd = 1;
options.max_M_iters = 1;
options.max_em_iters = 10;
options.em_tol = 1e-8;
[estLambda_z, estTheta_yz, estLambda_y, estTheta_xy, estStats] = em_scggm(...
    Z, Y_o, X, ...
    lambdaLambda_z, lambdaTheta_yz, lambdaLambda_y, lambdaTheta_xy, ...
    options);
estParams.xy = estTheta_xy;
estParams.yy = estLambda_y;
estParams.yz = estTheta_yz;
estParams.zz = estLambda_z;

figure('name', 'EM comparison');
subplot(3,4,4);
spy(Lambda_z); title('true \Lambda_z');
subplot(3,4,3);
spy(Theta_yz); title('true \Theta_{yz}');
subplot(3,4,2);
spy(Lambda_y); title('true \Lambda_y');
subplot(3,4,1);
spy(Theta_xy); title('true \Theta_xy');
subplot(3,4,8);
spy(estLambda_z); title('est \Lambda_z');
subplot(3,4,7);
spy(estTheta_yz); title('est \Theta_{yz}');
subplot(3,4,6);
spy(estLambda_y); title('est \Lambda_y');
subplot(3,4,5);
spy(estTheta_xy); title('est \Theta_{xy}');
